internal/zinc-core/src/main/scala/sbt/internal/inc/PickleJar.scala

```diff
@@ -13,201 +13,30 @@ package sbt
 import java.nio.file.attribute.{ BasicFileAttributes, FileTime }
 import java.time.Instant
-import java.util.Arrays
-import java.util.zip.{ CRC32, Deflater, ZipEntry, ZipOutputStream }
 
-import sbt.util.Logger
-import xsbti.compile.PickleData
-import scala.collection.JavaConverters._
-import scala.collection.mutable
+import scala.reflect.io.RootPath
 
 object PickleJar {
-  def write(
-      pickleOut: Path,
-      data: Iterable[PickleData],
-      knownProducts: collection.Set[String],
-      compilerVersion: String,
-      log: Logger
-  ): Path = {
-    // def trace(msg: String) = log.trace(() => new Exception(msg))
-    var pj: RootJarPath = null
-    val writtenPickles = new java.util.IdentityHashMap[AnyRef, String]()
-    val writtenSyms = new mutable.HashSet[String]
-    val sigWriter = FileWriter(pickleOut, compilerVersion, None)
+  def write(pickleOut: Path, knownProducts: Set[String]): Path = {
+    val pj = RootPath(pickleOut, writable = false) // so it doesn't delete the file
+    try Files.walkFileTree(pj.root, deleteUnknowns(knownProducts))
+    finally pj.close()
 
-    def writeSigFile(pickle: PickleData): Unit = {
-      val fqcn = pickle.fqcn()
-      // Reference to original scalac PickleBuffer (as AnyRef to avoid version dependence).
-      // For some reason, these might be duplicated.
-      val orig = pickle.underlying()
-      if (!writtenPickles.containsKey(orig)) {
-        if (writtenSyms.contains(fqcn))
-          log.warn(s"Found duplicate fqcn $fqcn while writing pickles!")
-        val elems: Iterable[String] = pickle.path.asScala.map(_.toString)
-        assert(elems.head == "__ROOT__")
-        val relativePath = elems.tail.mkString("/")
-        sigWriter.writeFile(relativePath, Arrays.copyOf(pickle.data, pickle.writeIndex))
-        writtenPickles.put(pickle.underlying, fqcn)
-        ()
-      } else ()
-    }
-
-    try {
-      for { pickle <- data } {
-        writeSigFile(pickle)
-      }
-    } finally {
-      sigWriter.close()
-    }
-    try {
-      pj = RootJarPath(pickleOut)
-      val knownPaths = knownProducts
-      Files.walkFileTree(
-        pj.root,
-        new SimpleFileVisitor[Path] {
-          override def visitFile(path: Path, attrs: BasicFileAttributes): FileVisitResult = {
-            val ps = path.toString
-            if (ps.endsWith(".sig")) {
-              // "/foo/bar/wiz.sig" -> "foo/bar/wiz.class"
-              val i0 = if (ps.startsWith("/")) 1 else 0
-              val cp = ps.substring(i0, ps.length - 3) + "class"
-              if (!knownPaths.contains(cp)) {
-                // log.info(s"deleting sig for removed class $path")
-                // trace(s"Removed $cp from picklejar")
-                Files.delete(path)
-              }
-
-            }
-            FileVisitResult.CONTINUE
-          }
-        }
-      )
-    } finally {
-      if (pj ne null) {
-        pj.close()
-      }
-    }
     Files.setLastModifiedTime(pickleOut, FileTime.from(Instant.now()))
```

Comment:
I wonder if this line is still needed?

Reply:
Never tried without, never looked into why it's there.  I'll push a commit dropping it.